vendor/github.com/containerd/containerd/grpc_services.go

```diff
@@ -0,0 +1,170 @@
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package containerd
+
+import (
+	"time"
+
+	containersapi "github.com/containerd/containerd/api/services/containers/v1"
+	contentapi "github.com/containerd/containerd/api/services/content/v1"
+	diffapi "github.com/containerd/containerd/api/services/diff/v1"
+	eventsapi "github.com/containerd/containerd/api/services/events/v1"
+	imagesapi "github.com/containerd/containerd/api/services/images/v1"
+	introspectionapi "github.com/containerd/containerd/api/services/introspection/v1"
+	leasesapi "github.com/containerd/containerd/api/services/leases/v1"
+	namespacesapi "github.com/containerd/containerd/api/services/namespaces/v1"
+	snapshotsapi "github.com/containerd/containerd/api/services/snapshots/v1"
+	"github.com/containerd/containerd/api/services/tasks/v1"
+	versionservice "github.com/containerd/containerd/api/services/version/v1"
+	"github.com/containerd/containerd/containers"
+	"github.com/containerd/containerd/content"
+	"github.com/containerd/containerd/dialer"
+	"github.com/containerd/containerd/images"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/containerd/containerd/snapshots"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/health/grpc_health_v1"
+)
+
+// grpcServices is an implementation of Services through grpc connection
+// with containerd grpc server.
+type grpcServices struct {
+	conn      *grpc.ClientConn
+	connector func() (*grpc.ClientConn, error)
+}
+
+func newGRPCServices(address string, copts clientOpts) (Services, error) {
+	gopts := []grpc.DialOption{
+		grpc.WithBlock(),
+		grpc.WithInsecure(),
+		grpc.WithTimeout(60 * time.Second),
+		grpc.FailOnNonTempDialError(true),
+		grpc.WithBackoffMaxDelay(3 * time.Second),
+		grpc.WithDialer(dialer.Dialer),
+	}
+	if len(copts.dialOptions) > 0 {
+		gopts = copts.dialOptions
+	}
+	if copts.defaultns != "" {
+		unary, stream := newNSInterceptors(copts.defaultns)
+		gopts = append(gopts,
+			grpc.WithUnaryInterceptor(unary),
+			grpc.WithStreamInterceptor(stream),
+		)
+	}
+	connector := func() (*grpc.ClientConn, error) {
+		conn, err := grpc.Dial(dialer.DialAddress(address), gopts...)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to dial %q", address)
+		}
+		return conn, nil
+	}
+	conn, err := connector()
+	if err != nil {
+		return nil, err
+	}
+	return &grpcServices{
+		conn:      conn,
+		connector: connector,
+	}, nil
+}
+
+func newGRPCServicesWithConn(conn *grpc.ClientConn) (Services, error) {
+	return &grpcServices{
+		conn: conn,
+	}, nil
+}
+
+// Reconnect re-establishes the GRPC connection to the containerd daemon
+func (g *grpcServices) Reconnect() error {
+	if g.connector == nil {
+		return errors.New("unable to reconnect to containerd, no connector available")
+	}
+	g.conn.Close()
```

Comment:
I see an error may be returned from Close() but the error check wasn't here in the original version. An error indicates close was already called. Not checking indicates we don't need to call close if it's already been called cause we did not care about the error. If it's already been called and we call close it cancels the context (presumably again, note context also uses a mutex sem and returns early if it's already cancelled), locks a mutex semaphore that protects the connection data, tests if conn.conns is nil, if it is nil it releases the mutex and returns ErrClientConnClosing. In short.. we don't care if it's called twice.

suggest adding an error check here if err != ErrClientConnClosing { return ("unable to reconnect, unexpected error while closing connection...

Reply:
This function will be empty for local services.

If you'd like you could send a PR to fix containerd if you think it needs a fix. :)